### PR TITLE
Bluetooth cache refactor

### PR DESCRIPTION
**Pull request by dati91**

### CODE REVIEW

File: tests/html/bluetooth_descriptor_info.html [hunk 1]
```diff
@@ -31,9 +31,9 @@
```
Comment (dati91) [inline tests/html/bluetooth_descriptor_info.html:31]: Maybe in a different commit?

File: components/net/bluetooth_thread.rs [hunk 2]
```diff
@@ -90,108 +93,353 @@ impl BluetoothManager {
...
-            appearance: Some(device.get_appearance()),
-            tx_power: Some(device.get_tx_power() as i8),
-            rssi: Some(device.get_rssi() as i8)
+            id: device.get_address().unwrap_or("".to_owned()),
+            name: device.get_name().ok(),
+            device_class: device.get_class().ok(),
+            vendor_id_source: device.get_vendor_id_source().ok(),
+            vendor_id: device.get_vendor_id().ok(),
+            product_id: device.get_product_id().ok(),
+            product_version: device.get_device_id().ok(),
```
Comment (zakorgy) [inline components/net/bluetooth_thread.rs:338]: get_device_id() gives back the product version?

Comment (dati91) [inline components/net/bluetooth_thread.rs:338]: Yes, but its called device id in the bluetooth api.

File: components/net/bluetooth_thread.rs [hunk 3]
```diff
@@ -21,7 +27,10 @@ pub trait BluetoothThreadFactory {
 impl BluetoothThreadFactory for IpcSender<BluetoothMethodMsg> {
     fn new() -> IpcSender<BluetoothMethodMsg> {
         let (sender, receiver) = ipc::channel().unwrap();
-        let adapter = BluetoothAdapter::init();
+        let adapter = match BluetoothAdapter::init() {
+            Ok(a) => Some(a),
+            Err(_) => None,
+        };
         spawn_named("BluetoothThread".to_owned(), move || {
             BluetoothManager::new(receiver, adapter).start();
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs:43]: What if adapter is None?

File: components/net/bluetooth_thread.rs [hunk 4]
```diff
@@ -90,108 +99,333 @@ impl BluetoothManager {
...
+            self.cached_devices.insert(device.get_address().unwrap_or("".to_owned()), device.clone());
+        }
+        devices
+    }
+
+    fn get_device(&mut self, adapter: &mut BluetoothAdapter, device_id: &str) -> Option<&BluetoothDevice> {
+        if self.cached_devices.contains_key(device_id) {
+            return self.cached_devices.get(device_id);
+        }
+        //update cache
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: style: `// Update`

File: components/net/bluetooth_thread.rs [hunk 5]
```diff
@@ -90,108 +99,333 @@ impl BluetoothManager {
...
+                        service_id: &str)
+                        -> Option<&BluetoothGATTService> {
+        if self.cached_services.contains_key(service_id) {
+            return self.cached_services.get(service_id);
+        }
+        let device_id = match self.service_to_device.get_mut(service_id) {
+            Some(d) => d.clone(),
+            None => return None,
+        };
+        //update cache
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: style

File: components/net/bluetooth_thread.rs (hunk 3 above)
Comment (dati91) [inline components/net/bluetooth_thread.rs:43]: then its None

File: components/net/bluetooth_thread.rs [hunk 6]
```diff
@@ -90,108 +99,325 @@ impl BluetoothManager {
...
+                                -> Option<BluetoothGATTService> {
+        for service in self.cached_services.values() {
+            if service.get_uuid().unwrap_or("".to_owned()) == *service_uuid {
+                return Some(service.clone());
+            }
+        }
+        // Update cache
+        let services = self.get_gatt_services(adapter, device_id);
+        for service in services {
+            if service.get_uuid().unwrap_or("".to_owned()) == *service_uuid {
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: Isn't there a function in the stdlib for this comparison?

File: components/net/bluetooth_thread.rs [hunk 7]
```diff
@@ -90,108 +99,325 @@ impl BluetoothManager {
...
-            None => {
-                sender.send(BluetoothObjectMsg::Error { error: String::from("No device found") }).unwrap();
-                return;
+        let connected = match self.get_device(&mut adapter, &device_id) {
+            Some(d) => {
+                if d.is_connected().unwrap_or(false) {
+                    d.disconnect().is_err()
+                } else {
+                    false
+                }
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs:387]: Can we optimize this else out?

**Pull request merged**